services/debugger/cache/embeddedcache/embeddedCache.go

```diff
@@ -0,0 +1,199 @@
+	cleanupFreq  time.Duration // TTL time on badgerDB
+	once         sync.Once
+	db           *badger.DB
+	Logger       logger.Logger
+}
+
+type badgerLogger struct {
+	logger.Logger
+}
+
+func (badgerLogger) Errorf(format string, a ...interface{}) {
+	_, _ = fmt.Fprintf(os.Stderr, format, a...)
+}
+
+func (badgerLogger) Warningf(format string, a ...interface{}) {
+	_, _ = fmt.Fprintf(os.Stderr, format, a...)
+}
+
+// Update writes the entries into badger db with a TTL
+func (e *EmbeddedCache[E]) Update(key string, value E) error {
+	e.Init()
+	txn := e.db.NewTransaction(true)
+	res := e.Read(key)
+	if len(res) >= e.limiter {
+		res = res[len(res)-e.limiter+1:]
+	}
+	res = append(res, value)
+	data, err := json.Marshal(res)
+	if err != nil {
+		return err
+	}
+	setBadgerEntry := func(ctx context.Context) error {
+		entry := badger.NewEntry([]byte(key), data).WithTTL(e.cleanupFreq)
+		if err := txn.SetEntry(entry); err == badger.ErrTxnTooBig {
+			err = txn.Commit()
+			if err != nil {
+				return err
+			}
+			txn = e.db.NewTransaction(true)
+			err = txn.SetEntry(entry)
+			if err != nil {
+				return err
+			}
+		}
+		return nil
+	}
+	err = misc.RetryWith(context.TODO(), e.queryTimeout, e.retries, setBadgerEntry)
+	if err != nil {
+		return err
+	}
+	err = txn.Commit()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// Read fetches all the entries for a given key from badgerDB
+func (e *EmbeddedCache[E]) Read(key string) []E {
+	e.Init()
+	read := func(ctx context.Context) ([]E, error) {
+		var values []E
+		err := e.db.View(func(txn *badger.Txn) error {
+			it := txn.NewIterator(badger.DefaultIteratorOptions)
+			defer it.Close()
+			prefix := []byte(key)
+			for it.Seek(prefix); it.ValidForPrefix(prefix); it.Next() {
+				item := it.Item()
+				err := item.Value(func(v []byte) error {
+					err := json.Unmarshal(v, &values)
+					if err != nil {
+						return err
+					}
+					return nil
+				})
+				if err != nil {
+					return err
+				}
+			}
+			return nil
+		})
+		if err != nil {
+			return nil, err
+		}
+		return values, nil
+	}
+
+	res, _ := misc.QueryWithRetries(context.TODO(), e.queryTimeout, e.retries, read)
+
+	return res
+}
+
+func (e *EmbeddedCache[E]) Init() {
+	e.once.Do(func() {
+		e.loadCacheConfig()
+		badgerPathName := e.Origin + "/cache/badgerdbv3"
+		tmpDirPath, err := misc.CreateTMPDIR()
+		if err != nil {
+			e.Logger.Errorf("Unable to create tmp directory: %v", err)
+			return
+		}
+		storagePath := path.Join(tmpDirPath, badgerPathName)
+		e.path = storagePath
+		e.done = make(chan struct{})
+		e.closed = make(chan struct{})
+		opts := badger.
+			DefaultOptions(storagePath).
+			WithLogger(badgerLogger{e.Logger}).
+			WithCompression(options.None).
+			WithIndexCacheSize(16 << 20). // 16mb
+			WithNumGoroutines(1).
+			WithNumMemtables(0).
+			WithBlockCacheSize(0)
+
+		e.db, err = badger.Open(opts)
+		if err != nil {
+			e.Logger.Errorf("Error while opening badgerDB: %v", err)
+			return
+		}
```

Comment:
We shouldn't start rudder-server in case of an error. We cannot ignore this and we shouldn't panic either:

1. `cache.New` should return an error along with `Cache[T]` (`EmbeddedCache` needs a proper producer function too and move away from this call-onceinit-in-every-method pattern)
2. The various debbuger (source, destination, transformation) `Start` methods should also propagate this error

Reply:
I don't agree with this actually... why. shouldn't we start rudder-server just because we are unable to initialise the cache? All of the errors in the live events module are actually recovered in a defer recover function. Cache is just an auxiliary functionality and should affect any of the existing pipeline behaviour right

Reply:
The various debugger modules (source, destination, transformation)  should propagate initialization errors. Otherwise they are hiding that they will be malfunctioning. No, design-wise logging an error is not enough. If you wish to ignore such initialization errors in the apphandlers then do it there.

Reply:
Refactored the code to propagate the errors to the top